handlertype.go

```diff
@@ -0,0 +1,257 @@
+		panic("unexpected handler type: " + t.String())
+	}
+}
+
+// MustNotBe panics if t is one of the given types.
+func (t HandlerType) MustNotBe(types ...HandlerType) {
+	if t.Is(types...) {
+		panic("unexpected handler type: " + t.String())
+	}
+}
+
+// IsConsumerOf returns true if handlers of type t can consume messages with the
+// given role.
+func (t HandlerType) IsConsumerOf(r message.Role) bool {
+	return r.Is(t.Consumes()...)
+}
+
+// IsProducerOf returns true if handlers of type t can produce messages with the
+// given role.
+func (t HandlerType) IsProducerOf(r message.Role) bool {
+	return r.Is(t.Produces()...)
+}
+
+// Consumes returns the roles of messages that can be consumed by handlers of
+// this type.
+func (t HandlerType) Consumes() []message.Role {
+	t.MustValidate()
+
+	switch t {
+	case AggregateHandlerType:
+		return []message.Role{message.CommandRole}
+	case ProcessHandlerType:
+		return []message.Role{message.EventRole, message.TimeoutRole}
+	case IntegrationHandlerType:
+		return []message.Role{message.CommandRole}
+	default: // ProjectionHandlerType
+		return []message.Role{message.EventRole}
+	}
+}
+
+// Produces returns the roles of messages that can be produced by handlers of
+// this type.
+func (t HandlerType) Produces() []message.Role {
+	t.MustValidate()
+
+	switch t {
+	case AggregateHandlerType:
+		return []message.Role{message.EventRole}
+	case ProcessHandlerType:
+		return []message.Role{message.CommandRole, message.TimeoutRole}
+	case IntegrationHandlerType:
+		return []message.Role{message.EventRole}
+	default: // ProjectionHandlerType
+		return nil
+	}
+}
+
+// ShortString returns a short (3-character) representation of the handler type.
+func (t HandlerType) ShortString() string {
+	t.MustValidate()
+
+	switch t {
+	case AggregateHandlerType:
+		return aggregateHandlerTypeShortString
+	case ProcessHandlerType:
+		return processHandlerTypeShortString
+	case IntegrationHandlerType:
+		return integrationHandlerTypeShortString
+	default: // ProjectionHandlerType
+		return projectionHandlerTypeShortString
+	}
+}
+
+// String returns a string representation of the handler type.
+func (t HandlerType) String() string {
+	switch t {
+	case AggregateHandlerType:
+		return aggregateHandlerTypeString
+	case ProcessHandlerType:
+		return processHandlerTypeString
+	case IntegrationHandlerType:
+		return integrationHandlerTypeString
+	case ProjectionHandlerType:
+		return projectionHandlerTypeString
+	default:
+		return fmt.Sprintf("<invalid handler type %#v>", t)
+	}
+}
+
+// MarshalText returns a UTF-8 representation of the handler type.
+func (t HandlerType) MarshalText() ([]byte, error) {
+	if err := t.Validate(); err != nil {
+		return nil, err
+	}
+
+	switch t {
+	case AggregateHandlerType:
+		return aggregateHandlerTypeBytes, nil
+	case ProcessHandlerType:
+		return processHandlerTypeBytes, nil
+	case IntegrationHandlerType:
+		return integrationHandlerTypeBytes, nil
+	default: // ProjectionHandlerType
```

Comment:
I do this where the value is pre-validated (as it is at the top of this function) to avoid having reams of uncovered `default:` blocks around the place.  I really wish Go had real enums with exhaustive checked switch statements.